### ISSUE TITLE
Argument '...' is defined as EntitySelector

### ISSUE THREAD

**Issue by marcingold** [JorelAli/1.13-Command-API#107]
**CommandAPI version**
3.1

**Minecraft version**
1.16.1

**Describe the bug**
When i create a command where the first argument is an new EntitySelectorArgument(EntitySelectorArgument.EntitySelector.MANY_PLAYERS)) argument, followed by a Custom Argument, i get the following error while try to tabcomplete the player / custom argument. Important to know, it looks like the command still executes correctly if u send it, even the autocomplete fails in that case
```log
[18:31:50] [Server thread/FATAL]: Error executing task on Server
java.lang.IllegalArgumentException: Argument 'players' is defined as EntitySelector, not interface java.util.Collection
        at com.mojang.brigadier.context.CommandContext.getArgument(CommandContext.java:92) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at dev.jorel.commandapi.CommandAPIHandler.lambda$getRequiredArgumentBuilderDynamic$9(CommandAPIHandler.java:640) ~[?:?]
        at com.mojang.brigadier.tree.ArgumentCommandNode.listSuggestions(ArgumentCommandNode.java:71) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at com.mojang.brigadier.CommandDispatcher.getCompletionSuggestions(CommandDispatcher.java:596) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at com.mojang.brigadier.CommandDispatcher.getCompletionSuggestions(CommandDispatcher.java:579) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.PlayerConnection.a(PlayerConnection.java:536) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.PacketPlayInTabComplete.a(SourceFile:35) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.PacketPlayInTabComplete.a(SourceFile:9) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:19) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.TickTask.run(SourceFile:18) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.executeTask(SourceFile:144) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.executeNext(SourceFile:118) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.MinecraftServer.aZ(MinecraftServer.java:943) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.MinecraftServer.executeNext(MinecraftServer.java:936) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.awaitTasks(SourceFile:127) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.MinecraftServer.sleepForTick(MinecraftServer.java:920) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.MinecraftServer.v(MinecraftServer.java:852) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at net.minecraft.server.v1_16_R1.MinecraftServer.lambda$0(MinecraftServer.java:164) ~[spigot-1.16.1.jar:git-Spigot-4d9262c-dc7c3c6]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]

```

**My code**
This code reproduces the behavior:
```java
    public Argument world() {
        return new CustomArgument<World>((input) -> {
            World world = Bukkit.getWorld(input);

            if (world == null) {
                throw new CustomArgument.CustomArgumentException(new CustomArgument.MessageBuilder("Unknown world: ").appendArgInput());
            }
            return world;
        }).overrideSuggestions((sender) -> this.plugin.getServer().getWorlds().stream().map(World::getName).toArray(String[]::new));
    }
```
```java
        CommandAPI.unregister("enchant", true);
        LinkedHashMap<String, Argument> enchantArguments = new LinkedHashMap<>();
        enchantArguments.put("players", new EntitySelectorArgument(EntitySelectorArgument.EntitySelector.MANY_PLAYERS));
        enchantArguments.put("enchantment", this.world());
        enchantArguments.put("level", new IntegerArgument());

        new CommandAPICommand("enchant")
                .withArguments(enchantArguments)                     // The arguments
                .withPermission(CommandPermission.OP)
                .executes((sender, args) -> {

                    World enchantment = (World) args[1];
                    int level = (int) args[2];

                    sender.sendMessage(enchantment.getName() + " on " + level);
                }).register();
});
```

**Expected behavior**
Autocomplete the player / custom Argument without error.

**Screenshots**
-

**Additional context**
I tried to strip down the example so u can just copy paste the code to any project to reproduce the error. So just ignore the naming of the variables ;)

**Comment by marcingold** [comment]
It looks like the context.getArgument call is the problem...

decompiled brigadier source where the error got thrown
```java
    public <V> V getArgument(String name, Class<V> clazz) {
        ParsedArgument<S, ?> argument = (ParsedArgument)this.arguments.get(name);
        if (argument == null) {
            throw new IllegalArgumentException("No such argument '" + name + "' exists on this command");
        } else {
            Object result = argument.getResult();
            if (((Class)PRIMITIVE_TO_WRAPPER.getOrDefault(clazz, clazz)).isAssignableFrom(result.getClass())) {
                return result;
            } else {
                throw new IllegalArgumentException("Argument '" + name + "' is defined as " + result.getClass().getSimpleName() + ", not " + clazz);
            }
        }
    }
```

Source of the error in the decompiled CommandAPI with Intelij (CommandAPIHandler#464)
```java
    private <T> RequiredArgumentBuilder<?, T> getRequiredArgumentBuilderDynamic(LinkedHashMap<String, Argument> args, String argumentName, Argument type, CommandPermission permission) {
        return type.getOverriddenSuggestions() == null ? (RequiredArgumentBuilder)RequiredArgumentBuilder.argument(argumentName, type.getRawType()).requires((clw) -> {
            return this.permissionCheck(nms.getCommandSenderForCLW(clw), permission);
        }) : this.getRequiredArgumentBuilderWithProvider(argumentName, type.getRawType(), permission, (context, builder) -> {
            List<Object> previousArguments = new ArrayList();
            Iterator var7 = args.keySet().iterator();

            while(var7.hasNext()) {
                String s = (String)var7.next();
                if (s.equals(argumentName)) {
                    break;
                }

                previousArguments.add(context.getArgument(s, ((Argument)args.get(s)).getPrimitiveType())); // error thrown from here
            }

            return this.getSuggestionsBuilder(builder, (String[])type.getOverriddenSuggestions().apply(nms.getCommandSenderForCLW(context.getSource()), previousArguments.toArray()));
        });
    }
```

Source of the error in the actual Command API Source (CommandAPIHandler#640)
```java
	// Gets a RequiredArgumentBuilder for a DynamicSuggestedStringArgument
	private <T> RequiredArgumentBuilder<?, T> getRequiredArgumentBuilderDynamic(
			final LinkedHashMap<String, Argument> args, String argumentName, Argument type,
			CommandPermission permission) {

		// If there are no changes to the default suggestions, return it as normal
		if (type.getOverriddenSuggestions() == null) {
			return RequiredArgumentBuilder.argument(argumentName, (ArgumentType<T>) type.getRawType())
					.requires(clw -> permissionCheck(nms.getCommandSenderForCLW(clw), permission));
		}

		// Otherwise, we have to handle arguments of the form BiFunction<CommandSender,
		// Object[], String[]>
		else {
			return getRequiredArgumentBuilderWithProvider(argumentName, type.getRawType(), permission,
					(CommandContext context, SuggestionsBuilder builder) -> {
						// Populate Object[], which is our previously filled arguments
						List<Object> previousArguments = new ArrayList<>();

						for (String s : args.keySet()) {
							if (s.equals(argumentName)) {
								break;
							}
							previousArguments.add(context.getArgument(s, args.get(s).getPrimitiveType())); // error thrown from here 
						}
						return getSuggestionsBuilder(builder, type.getOverriddenSuggestions()
								.apply(nms.getCommandSenderForCLW(context.getSource()), previousArguments.toArray()));
					});
		}
	}
```

I hope this will help you out in some way ;)

**Comment by JorelAli** [comment]
Going ahead and decompiling random segments of code was completely unnecessary and unhelpful. As a Java developer who has managed to create this project, which is a very complex system, all I needed was the error log and a detailed description of the bug. I am fully aware how to read Java error logs and I felt that your input was shadowing my knowledge.

Decompiling my code, was completely uncalled for - especially considering the source code is already present online. Your decompilation does not aid developers - it confuses them even more because decompilation is not a perfect operation. For example, in your decompiled code it had the line `Iterator var7 = args.keySet().iterator();`. Nowhere in my code have I ever written such a line. I'd recommend not doing this in future.

-----

That aside, I have resolved the issue in this commit: 53bbf39db154bdb042002e2888b5447f6f763cc7. I will include this in the 3.2 release.

**Comment by marcingold** [comment]
First of all thanks for fixing the bug :)

I never intended to question your knowledge.The decompiled code is just the way I found the affected place in the CommandAPI (hadn't cloned the repo at that time). Since I use Intelij, I could easily get to the decompiled code segment in the terminal by simply pressing "CTRL+Click". After that I searched for the corresponding line in the CommandAPI. I know that this is definitely not the cleanest way to find an error but for me it was just easier at that point. I just thought it wouldn't hurt anyone if I added what I found.  

But I think you're right and it might look a bit weird if someone comes across this issue and finds decompiled code . So I will delete the comment with the decompiled code, since it didn't help to fix the bug anyway

Greetings
Marc
